sphinx/transforms/post_transforms/__init__.py

```diff
@@ -197,11 +200,71 @@ def apply(self):
                     node.replace_self(nodes.comment())
 
 
+class SphinxSmartQuotes(SmartQuotes, SphinxTransform):
+    """
+    Customized SmartQuotes to avoid transform for some extra node types.
+    """
+    def apply(self):
+        # type: () -> None
+        if self.is_available():
+            self.env.settings['smart_quotes'] = True
```

Comment:
This should override `self.document.settings` instead. docutils' SmartQuotes class refers it.

`env.settings` is not referred here. It is prepared on `BuildEnvironment` class and used for create document object (see `sphinx.io:read_doc()`). So modifying the key is too late here.